충남대 BE_이은경 4주차 과제 (1단계)

pkyung opened the pull request:
## 과제 진행 사항
### 상품 카테고리
- [x] 카테고리 도메인 설계
- [x] 상품과 관계 설정
- [x] 카테고리 ru 구현
- [x] 화면 반영 - 상품과 함께 추가

## 구현에 대한 부가 설명
카테고리는 상품 생성 이전에 데이터베이스에 미리 들어가 있지 않으면 오히려 이상한 프로그램이라는 생각이 들어서 카테고리는 이미 db에 입력되었다고 생각하고 상품과 카테고리 구현을 진행했습니다.

pkyung committed:
Chore: 코드 옮기기 - step0

pkyung committed:
Docs: 기능 목록 작성 - step1

pkyung committed:
Feat: 카테고리 도메인 설계

카테고리 엔티티 설계를 구현했다. category의 필드는 id, name, color,
imageUrl, description으로 이뤄져 있으며 모든 필드는 null이 불가하도록
설정했다.

pkyung committed:
Feat: 카테고리의 관계 설정

카테고리는 상품과 1대1 관계라고 생각했다. 따라서, Product 도메인에
@OneToOne 관계로 설정을 해주었다. Category에는 아직 상품을 확인할 필요가
없을 것 같아서 양방향 매핑을 하지 않았다.

pkyung committed:
Feat: 카테고리 ru 구현

카테고리는 상품과 함께 지정할 수 있다는 요구사항이 있다. 따라서 Product
생성자와 update 를 category를 넣은 채로 수정했다.
그리고 category는 수정이 가능하고 모든 category를 조회하는 api가
요구사항에 존재한다. 따라서, findAllCategories와 updateCategory 메서드를
만들어서 카테고리를 조회하고 수정하도록 했다.

pkyung committed:
Refactor: handler 클래스 정리

notfound, alreadyExists, MethodArgument, Unauthorized exception 들을
다루는 handler로 클래스를 정리했다.

pkyung committed:
Chore: 안 옮긴 test 코드 옮김

pkyung committed:
Fix: 상품에 카테고리는 필수로 들어가야 해서 수정함

pkyung committed:
Fix: @ManyToOne으로 수정함

pkyung committed:
Feat: 화면 반영 - 상품과 함께 카테고리  추가되도록

Product를 Page<Product>로 반환하면 Product만 반환되고 연관관계 설정된
Category는 나오지 않아서 ProductResponse를 생성하여 반환하도록 수정헸다.
Product와 Category에 필요한 getter들을 생성했다.

pkyung committed:
Refactor: 필요없는 import 삭제

pkyung committed:
Refactor: 하나만 조회하는 api 또한 productResponse 반환

pkyung committed:
Refactor: token 관련 메서드 import 해서 사용

pkyung committed:
Fix: wish.js의 createWish 전역 변수 설정

pkyung committed:
Refactor: category getter 추가

pkyung committed:
Test: 상품과 카테고리 저장 테스트

pkyung committed:
Docs: 기능 목록 구현 - step1

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Refactor: merge 시 삭제 안된 파일 삭제

ksy90101 commented on src/main/java/gift/controller/CategoryController.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.List;
+
+@RequestMapping("/api")
```
여기에 categories를 같이 적으면 좋지않을까요?

pkyung replied:
네 수정하겠습니다.

ksy90101 commented on src/main/java/gift/controller/CategoryController.java:
```diff
@@ -0,0 +1,35 @@
+
+    @GetMapping("/categories")
+    public ResponseEntity findAllCategories() {
+        List<Category> categoryList = categoryService.findAllCategories();
```
콜렉션인 경우 변수명에 콜렉션을 쓰기보딘 복수형으로 적는게 더 명확할수 있습니다.



참고: https://devlog.jsyoo5b.net/ko/posts/clean-code/ch02/

> 예를 들어 accountList는 해당 변수가 진짜 List 형식이 아니라면 사용하지 말고, 차라리 accountGroup, bunchOfAccounts, 혹은 그냥 accounts로 표현하는 것이 더 낫다고 한다.

pkyung replied:
참고하고 수정하겠습니다!

ksy90101 commented on src/main/java/gift/controller/CategoryController.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @PutMapping("/categories")
+    public ResponseEntity updateCategory(CategoryRequest request) {
```
컨트롤러를 보면 카테고리라는걸 이해할수 있으니. Update만 적어도 되지않을까요?

ksy90101 replied:
따로 body가 없기 때문에 Void로 제네릭의 타입을 지정하면 더 의미전달이 잘 될거 같아요~

pkyung committed:
Refactor: 불필요한 import 제거

pkyung committed:
Refactor: merge 시 삭제 안된 파일 삭제

ksy90101 commented on src/main/java/gift/controller/CategoryController.java:
```diff
@@ -0,0 +1,35 @@
+    @PutMapping("/categories")
+    public ResponseEntity updateCategory(CategoryRequest request) {
+        categoryService.updateCategory(request);
+        return ResponseEntity.status(HttpStatus.CREATED).build();
```
201 created는 기본적으로 의미가 리소스가 생성될때를 의미합니다.
따라서 update는 생성되었다는 의미는 아니기 때문에 

200이나 지금처럼 내려줄 body가 없다면 204 no_content를 쓰는 편입니다.

참고링크 : https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204

pkyung replied:
update 할 때의 상태 반환을 어떻게 할 지에 대해 잘 몰랐는데 감사합니다. 수정하겠습니다.

ksy90101 commented on src/main/java/gift/dto/CategoryRequest.java:
```diff
@@ -0,0 +1,16 @@
+        Long id,
+        @NotBlank
+        String name,
+        @NotBlank
+        String color,
+        @NotBlank
+        String imageUrl,
```
여기도 color코드와 url validate를 하면 좋을거 같습니다.

pkyung replied:
정규식을 통해 검증을 진행해보았습니다.

ksy90101 commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -10,7 +11,11 @@ public record ProductRequest(Long id,
                              int price,
-                             String imageUrl
+                             @NotBlank
+                             String imageUrl,
```
여기도 url validate가 필요해보입니다.

ksy90101 commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -10,7 +11,11 @@ public record ProductRequest(Long id,
                              @Pattern(regexp = "[a-zA-Z0-9가-힣\\(\\)\\[\\]\\-+&_\\/\\s]+", message = "상품 이름에는 (), [], -, +, &, _, /, 공백을 제외한 특수 문자를 사용할 수 없습니다.")
                              @Pattern(regexp = "^(?!.*카카오).*$", message = "상품 이름에 '카카오' 가 포함 되어 있습니다. 담당 MD와 협의가 필요합니다.")
                              String name,
+                             @NotNull
                              int price,
```
boxed Type이 아닌경우 "int"에서 null이 들어올수 있을까요?

참고링크: https://unit-15.tistory.com/172

pkyung replied:
수정했습니다.

ksy90101 commented on src/main/java/gift/dto/ProductResponse.java:
```diff
@@ -0,0 +1,11 @@
+        String name,
+        int price,
+        String imageUrl,
+        String categoryName,
```
지금도 괜찮지만, 대부분 계층형으로 두기도 합니다.

{
   id:
....
  category: {
     id:
     name: 
  }
}

이렇게 하면 훨씬 모델을 이해하는게 쉽거든요~

pkyung replied:
계층형으로 수정하겠습니다!

ksy90101 commented on src/main/java/gift/exception/NotFoundExceptionHandler.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class NotFoundExceptionHandler {
```
점점 NotFound가 많아지는거 같아요.
이런경우 NotFoundException을 만들어서 공통으로 처리하기도 합니다.

모델이 만들어질때마다 Exception을 만들면 힘드니깐요.

cf) MemberNotFoundException은 다른 로직이기 떄문에 공통으로 하기 어려울거 같아요 ㅎㅎ 이 부분은 뺴고 ㅎㅎ

pkyung replied:
이전의 피드백을 보고 Exception Handler를 notfound, alreadyExists 등으로 모은 것인데 notfound과 같은 Exception을 공통으로 처리하라는 말씀이었군요,,, 혹시 그러면 Handler 또한 다시 분리해야할까요?

pkyung replied:
exception을 하나로 모은 다음 global handler를 두는 게 나을까요?

ksy90101 replied:
분리에 대해서는 정답이 존재하지 않습니다.

저같은 경우엔 공통인 경우엔 GlobalHendler에 넣고, 각 컨트롤러에서 각각의 예외 핸들러가 필요하면 그때가서 예외에 대한 컨트롤러를 넣는 편입니다.

ksy90101 commented on src/main/java/gift/model/Category.java:
```diff
@@ -0,0 +1,63 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String name;
```
요구사항에서 name은 유니크한값으로 봤는데.. ㅎㅎ 유니크 인덱스 처리가 안되어있네요!?

pkyung replied:
요구사항에 힌트가 추가되어 확인하지 못했습니다.  추가해서 수정하겠습니다!

ksy90101 commented on src/main/java/gift/model/Category.java:
```diff
@@ -0,0 +1,63 @@
+    private Long id;
+    @Column(nullable = false)
+    private String name;
+    @Column(nullable = false)
+    private String color;
+    @Column(nullable = false)
+    private String imageUrl;
```
이번엔 한번 모델에도 validate를 고려해보죠.
엔티티, 모델에도 validate가 필요합니다.

한번 validate를 해보시길 바랍니다.

pkyung replied:
고려해서 수정해보겠습니다.

ksy90101 commented on src/main/java/gift/model/Category.java:
```diff
@@ -0,0 +1,63 @@
+    private String color;
+    @Column(nullable = false)
+    private String imageUrl;
+    @Column(nullable = false)
+    private String description;
```
description이 255자로 충분할까요?

"@Lob"을 적용할 필요는 없을까요?

pkyung replied:
적용해야할 것 같습니다. 추가하겠습니다!

ksy90101 commented on src/main/java/gift/service/CategoryService.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Transactional
+    public Category updateCategory(CategoryRequest request) {
+        Optional<Category> category = categoryRepository.findById(request.id());
```
여기서 orElesThrow로 추출해서 처리하면 더 편하지 않을까요?

pkyung replied:
수정하겠습니다!

ksy90101 reviewed (changes requested):
간단한 피드백 남겼습니다~
확인 부탁드려요!
고생많으셨습니다!
질문이 있다면 언제든 DM주세요!

pkyung committed:
Refactor: category controller와 service 리펙토링

1. requestMapping에 /api/categories 로 수정했다.
2. 메서드명 updateCategory -> update
3. category list의 변수병을 복수형으로 수정했다.
4. update 의 경우 반환할 때 no_content
5. body가 없는 responseEntity는 Void 제네릭 타입으로 명시했다.

pkyung committed:
Refactor: request validation 추가

1. ProductRequest와 CategoryRequest의 color 필드를 정규식으로 검증했다.
2. imageUrl 필드를 정규식을 통해 검증했다.

pkyung committed:
Refactor: productResponse 계층형으로 수정

1. categoryName, categoryId를 category 객체로 수정했다.
2. service의 반환값의 생성자 일부를 수정했다.

pkyung committed:
Refactor: category name 유니크 추가

pkyung committed:
Refactor: orElseThrow로 optional 뗌

pkyung committed:
Refactor: category 엔티티 validate 설정 추가

pkyung committed:
Refactor: description 필드 255자 이상을 위해 @Lob 추가

pkyung committed:
Refactor: Exception 관리

1. AlreadyExistsException과 NotFoundException 을 생성했다.
2. Service에 사용된 Exception들을 수정했다.
3. GlobalExceptionHandler에서 exception을 관리하도록 수정했다.

pkyung committed:
Refactor: 메서드 이름 수정

ksy90101 commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -13,51 +16,79 @@
-            Product updateProduct = optionalProduct.get().update(
-                    request.name(),
-                    request.price(),
-                    request.imageUrl()
-            );
-            return updateProduct;
+        if (!optionalProduct.isPresent()) {
+            throw new NotFoundException("수정하려는 해당 id의 상품이 존재하지 않습니다.");
+        }
+        if (!optionalCategory.isPresent()) {
+            throw new NotFoundException("해당 카테고리가 존재하지 않습니다.");
```
Optional을 쓰는곳을 모두 orElseThrow 형식으로 변경하면 어떠까요?

ksy90101 reviewed (approved):
고생하셨어요!
Optional의 메서드 체이닝에 대해서 좀더 찾아보시면 좋을거 같습니다.
2단계에서 뵐게요~
